components/places/src/storage/bookmarks.rs

```diff
@@ -786,6 +786,40 @@ pub fn bookmarks_get_url_for_keyword(db: &PlacesDb, keyword: &str) -> Result<Opt
     }
 }
 
+fn repeat_vars(count: usize) -> String {
+    assert_ne!(count, 0);
+    let mut s = "?,".repeat(count);
+    // Remove trailing comma
+    s.pop();
+    s
+}
+
+// Counts the number of bookmark items in the bookmark trees under the specified GUIDs.
+// Does not count folder items, separators. A set of empty folders will return zero, as will
+// a set of non-existing GUIDs or guids of a non-folder item.
+// The result is implementation dependant if the trees overlap.
+pub fn count_bookmarks_in_trees(db: &PlacesDb, item_guids: &[SyncGuid]) -> Result<u32> {
+    if item_guids.is_empty() {
+        return Ok(0);
+    }
+    let vars = repeat_vars(item_guids.len());
+    let sql = format!(
+        r#"
+        WITH RECURSIVE bookmark_tree(id, parent, type)
+        AS (
+            SELECT id, parent, type FROM moz_bookmarks
+            WHERE parent IN (SELECT id from moz_bookmarks WHERE guid IN ({vars}))
+            UNION ALL
+            SELECT bm.id, bm.parent, bm.type FROM moz_bookmarks bm, bookmark_tree bt WHERE bm.parent = bt.id
+        )
+    SELECT COUNT(*) from bookmark_tree WHERE type = {};
+    "#,
+        BookmarkType::Bookmark as u8
+    );
+    let params = rusqlite::params_from_iter(item_guids);
+    Ok(db.query_row_and_then(&sql, params, |row| row.get(0))?)
```

Comment:
If the types work out, could we use our `db.query_one` extension from `sql_support` here?

Reply:
It's unfortunate that `db.query_one` doesn't take params :( try_query_one does, so there's odd asymmetry there - I moved to try_query_one anyway even though I don't think we could ever return 0 rows.